src/gebt_poc/solver.cpp

```diff
@@ -234,4 +234,303 @@ Kokkos::View<double*> CalculateStaticResidual(
+
+    // C12 consists of the (0, 3) -> (2, 5) components of the sectional stiffness matrix
+    auto C12 = Kokkos::View<double**>("C12", 3, 3);
+    Kokkos::parallel_for(
+        Kokkos::MDRangePolicy<Kokkos::DefaultExecutionSpace, Kokkos::Rank<2>>({0, 0}, {3, 3}),
+        KOKKOS_LAMBDA(const size_t i, const size_t j) {
+            C12(i, j) = sectional_stiffness(i, j + kNumberOfVectorComponents);
+        }
+    );
+
+    // C21 consists of the (3, 0) -> (5, 2) components of the sectional stiffness matrix
+    auto C21 = Kokkos::View<double**>("C21", 3, 3);
+    Kokkos::parallel_for(
+        Kokkos::MDRangePolicy<Kokkos::DefaultExecutionSpace, Kokkos::Rank<2>>({0, 0}, {3, 3}),
+        KOKKOS_LAMBDA(const size_t i, const size_t j) {
+            C21(i, j) = sectional_stiffness(i + kNumberOfVectorComponents, j);
+        }
+    );
+
+    // Calculate the two non-zero submatrices
+    auto x0_prime_tilde =
+        gen_alpha_solver::create_cross_product_matrix(gen_alpha_solver::create_vector(
+            {pos_vector_derivatives(0), pos_vector_derivatives(1), pos_vector_derivatives(2)}
+        ));
+    auto u_prime_tilde =
+        gen_alpha_solver::create_cross_product_matrix(gen_alpha_solver::create_vector(
+            {gen_coords_derivatives(0), gen_coords_derivatives(1), gen_coords_derivatives(2)}
+        ));
+    auto values = gen_alpha_solver::add_matrix_with_matrix(x0_prime_tilde, u_prime_tilde);
+
+    auto N_tilde = gen_alpha_solver::create_cross_product_matrix(N);
+    auto M_tilde = gen_alpha_solver::create_cross_product_matrix(M);
+
+    // Assemble the O matrix
+    // [O]_6x6 = [
+    //     [0]_3x3      -N_tilde + [C11] * (x_0_prime_tilde + u_prime_tilde)
+    //     [0]_3x3      -M_tilde + [C21] * (x_0_prime_tilde  + u_prime_tilde)
+    // ]
+
+    // Calculate the two non-zero submatrices
+    auto non_zero_terms_part_1 = gen_alpha_solver::add_matrix_with_matrix(
+        gen_alpha_solver::multiply_matrix_with_scalar(N_tilde, -1.),
+        gen_alpha_solver::multiply_matrix_with_matrix(C11, values)
+    );
+    auto non_zero_terms_part_2 = gen_alpha_solver::add_matrix_with_matrix(
+        gen_alpha_solver::multiply_matrix_with_scalar(M_tilde, -1.),
+        gen_alpha_solver::multiply_matrix_with_matrix(C21, values)
+    );
+
+    auto O_matrix =
+        Kokkos::View<double**>("O_matrix", kNumberOfLieGroupComponents, kNumberOfLieGroupComponents);
+    Kokkos::parallel_for(
+        Kokkos::MDRangePolicy<Kokkos::DefaultExecutionSpace, Kokkos::Rank<2>>(
+            {0, 0}, {kNumberOfVectorComponents, kNumberOfVectorComponents}
+        ),
+        KOKKOS_LAMBDA(const size_t i, const size_t j) {
+            O_matrix(i, j) = 0.;
+            O_matrix(i + kNumberOfVectorComponents, j) = 0.;
+            O_matrix(i, j + kNumberOfVectorComponents) = non_zero_terms_part_1(i, j);
+            O_matrix(i + kNumberOfVectorComponents, j + kNumberOfVectorComponents) =
+                non_zero_terms_part_2(i, j);
+        }
+    );
```

Comment:
Perhaps each of the O, P, and Q matrices should be assembled in their own function?  This would help with independent testing and make things a little more readable than a single big function.